core/integration/remotecache_test.go

```diff
@@ -0,0 +1,106 @@
+package core
+
+import (
+	"context"
+	"os"
+	"os/exec"
+	"testing"
+	"time"
+
+	"dagger.io/dagger"
+	"github.com/moby/buildkit/identity"
+	"github.com/stretchr/testify/require"
+)
+
+func TestRemoteCache(t *testing.T) {
+	// TODO: until this setting is configurable at runtime, just spawning separate engines w/ the config set
+	ctx, cancel := context.WithCancel(context.Background())
+	defer cancel()
+
+	registryContainerName := runRegistryInDocker(ctx, t)
+
+	getClient := func() *dagger.Client {
+		return runSeparateEngine(ctx, t, map[string]string{
+			"_EXPERIMENTAL_DAGGER_CACHE_CONFIG": "type=registry;ref=127.0.0.1:5000/test-cache",
+		}, "container:"+registryContainerName)
+	}
+	pipelineOutput := func(c *dagger.Client) string {
+		output, err := c.Container().From("alpine:3.17").WithExec([]string{
+			"sh", "-c", "head -c 128 /dev/random | sha256sum",
+		}).Stdout(ctx)
+		require.NoError(t, err)
+		return output
+	}
+
+	clientA := getClient()
+	clientB := getClient()
+	outputA := pipelineOutput(clientA)
+	require.NoError(t, clientA.Close()) // needed in order to end the session and push its cache
+	outputB := pipelineOutput(clientB)
+	require.Equal(t, outputA, outputB)
```

Comment:
:exploding_head: These 6 lines are so unassuming until you pay attention to what's going on.

1. Start a registry for storing the cache.
1. Start two independent Dagger engines with clients configured to use the registry as a cache.
1. Use client A to run a pipeline that generates a unique `stdout`.
1. Close the client, thereby ending the session and exporting its cache.
1. Use client B to run the same pipeline again, and **see the same result** due to the cache populated by the other session.

Cool stuff.

Reply:
I guess I should add a comment explaining this 😅